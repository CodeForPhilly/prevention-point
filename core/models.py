from django.db import models
from django.utils import timezone
from enum import Enum
from core.participants.models import Participant
class Medication(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    medical_delivery = models.CharField(max_length=100)
    medication_name = models.CharField(max_length=100)
    ingestion_frequency = models.IntegerField()

class UrineDrugScreen(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    date_of_test = models.DateField()
    uds_temp = models.CharField(max_length=100, verbose_name="Urine Temperature")
    # TODO: does this belong here? if so, is it a Bool for if a PT was taken, or Bool for if pregnant?
    pregnancy_test = models.BooleanField(default=False)
    opiates = models.BooleanField(default=False)
    fentanyl = models.BooleanField(default=False)
    bup = models.BooleanField(default=False, verbose_name="Buprenorphine")
    coc = models.BooleanField(default=False, verbose_name="Cocaine")
    amp = models.BooleanField(default=False, verbose_name = "Amphetamine")
    m_amp = models.BooleanField(default=False, verbose_name = "Methamphetamine")
    thc = models.BooleanField(default=False, verbose_name = "THC")
    mtd = models.BooleanField(default=False, verbose_name = "Methadone")
    pcp = models.BooleanField(default=False, verbose_name = "PCP")
    bar = models.BooleanField(default=False, verbose_name = "Barbiturates")
    bzo = models.BooleanField(default=False, verbose_name = "Benzodiazepines")
    tca = models.BooleanField(default=False, verbose_name = "Tricyclic Antidepressants")
    oxy = models.BooleanField(default=False, verbose_name = "Oxycodone")

    def __str__(self):
        return 'UDS #%i for Participant #%i' % (self.id, self.participant_id)

class EmployeeRole(models.Model):
    role_value = models.CharField(max_length=30)

    def __str__(self):
        return '%s' % (self.role_value)

# TODO: seed with front desk, case manager, admin

class Employee(models.Model):
    first_name = models.CharField(max_length=50)
    last_name = models.CharField(max_length=100)
    role = models.ForeignKey(EmployeeRole, on_delete=models.CASCADE)

    def __str__(self):
        return '%s %s %s' % (self.first_name, self.last_name, self.role)

class CaseManagement(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    crs_seen = models.BooleanField(default=False)

class Appointment(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)

    cm_employee = models.ForeignKey(
            Employee, related_name='cm_%(class)s', on_delete=models.CASCADE
            )

    crs_employee = models.ForeignKey(
            Employee, related_name='crs_%(class)s', on_delete=models.CASCADE
            )

    appointment_timestamp = models.DateTimeField()
    crs_timestamp = models.DateTimeField()
    crs_seen = models.BooleanField(default=False)
    cm_timestamp = models.DateTimeField()
    cm_seen = models.BooleanField(default=False)
    hcv_status = models.BooleanField(default=False)

# TODO: we probably want to move the HCV data into its own table(s), similar to how UDS works

class BehavioralHealthNotes(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    employee = models.ForeignKey(Employee, on_delete=models.CASCADE)
    note_timestamp = models.DateTimeField()
    behavior_note = models.TextField()

class HCVNotes(models.Model):
    participant = models.ForeignKey(Appointment, on_delete=models.CASCADE)
    employee = models.ForeignKey(Employee, on_delete=models.CASCADE)
    note_timestamp = models.DateTimeField()
    hcv_note = models.TextField()

class Visit(models.Model):
    participant = models.ForeignKey(Participant, on_delete=models.CASCADE)
    created_at = models.DateTimeField(default=timezone.now)

class ServiceEvent(models.Model):
    class ServiceEventPurpose(Enum):
        ARRIVED = 'arrived'
        CALLED = 'called'
        SEEN = 'seen'
        LEFT = 'left'

    visit = models.ForeignKey(Visit, on_delete=models.CASCADE)
    # service = models.ForeignKey(Service, on_delete=models.CASCADE)
    purpose = models.CharField(
        max_length=10,
        choices=[(tag.name, tag.value) for tag in ServiceEventPurpose]
    )
