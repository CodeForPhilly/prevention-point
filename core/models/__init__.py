from .insurer import *
from .participant import *
from .program import *
from .service import *
from .program_service_map import *
from .visit import *
from .hcv_notes import *
from .medication import *
from .appointment import *
from .case_management import *
from .front_desk_event import *
from .urine_drug_screen import *
from .program_availability import *
from .behavioral_health_notes import *