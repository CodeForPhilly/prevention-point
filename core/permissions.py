from django.contrib.auth.models import User, Group
from rest_framework.permissions import DjangoModelPermissions, BasePermission

FRONT_DESK = 'front_desk'
CASE_MANAGER = 'case_manager'
ADMIN = 'admin'

class DjangoModelPermissions(DjangoModelPermissions):
    """
    The request is authenticated using `django.contrib.auth` permissions.
    See: https://docs.djangoproject.com/en/dev/topics/auth/#permissions
    It ensures that the user is authenticated, and has the appropriate
    `add`/`change`/`delete` permissions on the model.
    This permission can only be applied against view classes that
    provide a `.queryset` attribute.


    Prevention Point Note: We are extending this Class as to include permission checks on GET
    requests. You MUST use this class to protect GET requests to endpoints. The DRF class this 
    ineherits from does not protect GET requests
    """
    perms_map = {
        'OPTIONS': [],
        'HEAD': [],
        'GET': ['%(app_label)s.view_%(model_name)s'],
        'POST': ['%(app_label)s.add_%(model_name)s'],
        'PUT': ['%(app_label)s.change_%(model_name)s'],
        'PATCH': ['%(app_label)s.change_%(model_name)s'],
        'DELETE': ['%(app_label)s.delete_%(model_name)s'],
    
    }
