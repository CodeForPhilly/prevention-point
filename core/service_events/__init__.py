from .serializer import *
from .views import *
