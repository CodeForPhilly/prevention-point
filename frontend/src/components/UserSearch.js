import React, { useState, useContext } from "react"
import { observer } from "mobx-react-lite"
import { useHistory } from "react-router-dom"

import Input from "@material-ui/core/Input"
import Container from "@material-ui/core/Container"
import FormGroup from "@material-ui/core/FormGroup"
import InputLabel from "@material-ui/core/InputLabel"
import FormControl from "@material-ui/core/FormControl"

import AppCopy from "./AppCopy"
import AppButton from "./AppButton"
import AppHeading from "./AppHeading"
import { rootStoreContext } from "../stores/RootStore"

const UserSearch = observer(() => {
  const rootStore = useContext(rootStoreContext)
  const participantsStore = rootStore.ParticipantStore
  const history = useHistory()

  const handleSubmit = e => {
    if (e) {
      e.preventDefault()
    }
    // Validate that a user id or a first name last name paring exist on the form
    if (!participantsStore.params) {
      participantsStore.setErrorMessageForUserSearch(
        "Need to enter a user id or a name"
      )
      participantsStore.setErrorMessageForUserSearch(true)
    } else {
      participantsStore.getParticipants()
      history.push("/participants")
    }
  }

  return (
    <Container className="participant-search">
      <div>
        <AppHeading className="participant-search__heading">
          Participant Search
        </AppHeading>
        <AppCopy className="participant-search__copy">
          <b>Reminder:</b> Search for participant profile prior to creating a
          new profile
        </AppCopy>
        <form className="participant-search__form" onSubmit={handleSubmit}>
          <FormGroup className="participant-search__input">
            <FormControl>
              <InputLabel htmlFor="user_id">User ID</InputLabel>
              <Input
                id="user_id"
                name="user_id"
                value={participantsStore.params.userId}
                onChange={e => participantsStore.setUserIdParam(e.target.value)}
                className={participantsStore.errorState ? "error" : ""}
              />
            </FormControl>
          </FormGroup>
          <AppHeading className="participant-search__heading">Or</AppHeading>
          <FormGroup className="participant-search__input">
            <FormControl>
              <InputLabel htmlFor="first_name">First Name</InputLabel>
              <Input
                id="first_name"
                name="first_name"
                value={participantsStore.params.firstName}
                onChange={e =>
                  participantsStore.setFirstNameParam(e.target.value)
                }
                className={participantsStore.errorState ? "error" : ""}
              />
            </FormControl>
          </FormGroup>
          <FormGroup className="participant-search__input">
            <FormControl>
              <InputLabel htmlFor="last_name">Last Name</InputLabel>
              <Input
                id="last_name"
                name="last_name"
                value={participantsStore.params.lastName}
                onChange={e =>
                  participantsStore.setLastNameParam(e.target.value)
                }
                className={participantsStore.errorState ? "error" : ""}
              />
            </FormControl>
          </FormGroup>
          <p className="error-message">{errorMessage}</p>
          <AppButton submit disabled={participantsStore.toggleSearch}>
            Submit
          </AppButton>
        </form>
      </div>
    </Container>
  )
})

export default UserSearch
