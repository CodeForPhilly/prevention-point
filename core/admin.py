from django.contrib import admin
from core.models import Participant, Medication, CaseManagement, Appointment, Employee, BehavioralHealthNotes
from core.service_events.models import ServiceEvent
from core.urine_drug_screens.models import UrineDrugScreen

# Register your models here.
admin.site.register(Medication)
admin.site.register(UrineDrugScreen)
admin.site.register(CaseManagement)
admin.site.register(Appointment)
admin.site.register(Employee)
admin.site.register(BehavioralHealthNotes)
admin.site.register(ServiceEvent)

class BaseInline(admin.StackedInline):
    extra = 0

class UrineDrugScreenInline(BaseInline):
    model = UrineDrugScreen

class AppointmentInline(BaseInline):
    model = Appointment

class MedicationInline(BaseInline):
    model = Medication

class CaseManagement(BaseInline):
    model = CaseManagement

class ParticipantAdmin(admin.ModelAdmin):
    inlines = [
            AppointmentInline,
            MedicationInline,
            CaseManagement,
            UrineDrugScreenInline,
            ]

admin.site.register(Participant, ParticipantAdmin)
