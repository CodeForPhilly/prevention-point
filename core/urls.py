"""preventionpoint URL Configuration

The `urlpatterns` list routes URLs to views. For more information please see:
    https://docs.djangoproject.com/en/2.1/topics/http/urls/
Examples:
Function views
    1. Add an import:  from my_app import views
    2. Add a URL to urlpatterns:  path('', views.home, name='home')
Class-based views
    1. Add an import:  from other_app.views import Home
    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
Including another URLconf
    1. Import the include() function: from django.urls import include, path
    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
"""
from django.contrib import admin
from django.urls import include, path
from rest_framework import routers
from rest_framework_swagger.views import get_swagger_view
from rest_framework_simplejwt import views as jwt_views

from core.users import views as user_views
from core.urine_drug_screens import views as uds_views
from core.participants import views as participant_views
from core.visits import views as visits_views
from core.front_desk_events import views as front_events_views
from core.programs import views as programs_views
from core.services import views as services_views
from core.queue import views as queues_views
from core.program_availability import views as program_availability_views
from core.program_service_map import views as program_service_map_views
from core.insurers import views as insurer_views
from core.medications import views as medications_views

admin.site.site_header = "Prevention Point Philadelphia"

router = routers.DefaultRouter()
router.register(r"users", user_views.UserViewSet)
router.register(r"groups", user_views.GroupViewSet)
router.register(r"visits", visits_views.VisitViewSet)
router.register(r"uds", uds_views.UrineDrugScreenViewSet)
router.register(r"medications", medications_views.MedicationViewSet)
router.register(r"participants", participant_views.ParticipantViewSet)
router.register(r"front-desk-events", front_events_views.FrontDeskEventViewSet)
router.register(r"programs", programs_views.ProgramViewSet)
router.register(r"program-service-map", program_service_map_views.ProgramServiceMapView)
router.register(r"services", services_views.ServiceViewSet)
router.register(r"insurers", insurer_views.InsurerViewSet)


schema_view = get_swagger_view(title="PreventionPoint API")

urlpatterns = [
    path("api/", include(router.urls)),
    path(
        "api/token/", jwt_views.TokenObtainPairView.as_view(), name="token_obtain_pair"
    ),
    path(
        "api/token/refresh/", jwt_views.TokenRefreshView.as_view(), name="token_refresh"
    ),
    path("api/token/verify/", jwt_views.TokenVerifyView.as_view(), name="token_verify"),
    path(
        "api/programs/<int:program_id>/program-availability/",
        program_availability_views.ProgramAvailabilityViewSet.as_view({"get": "list"}),
    ),
    path(
        "api/programs/<int:program_id>/program-availability/<int:pk>/",
        program_availability_views.ProgramAvailabilityViewSet.as_view(
            {"put": "update"}
        ),
    ),
    path(
        "api/programs/<int:program_id>/queue/",
        queues_views.QueueViewSet.as_view({"get": "retrieve"}),
    ),
    # above only 'registers' the retrieve path. to use other CRUD actions, must pass that explicitly to .as_view in another register
    path("swagger/", schema_view),
    path("admin/", admin.site.urls),
]
