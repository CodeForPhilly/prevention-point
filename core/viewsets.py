from rest_framework import generics, viewsets
from rest_framework.permissions import IsAuthenticated
from core.permissions import DjangoModelPermissions

class ModelViewSet(viewsets.ModelViewSet):
    permission_classes = [DjangoModelPermissions, IsAuthenticated]
