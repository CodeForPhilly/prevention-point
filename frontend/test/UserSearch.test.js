/* eslint-disable quotes */
import React from "react"
import { configure } from "enzyme"
import { createMount } from "@material-ui/core/test-utils"
import Adapter from "enzyme-adapter-react-16"
import { ThemeProvider, createMuiTheme } from "@material-ui/core/styles"

import UserSearch from "../src/components/UserSearch"

configure({ adapter: new Adapter() })

const theme = createMuiTheme({})

describe("<UserSearch />", () => {
  // eslint-disable-next-line no-unused-vars
  let wrapper
  //let shallow
  let mount
  let buttons

  const initialProps = {}

  beforeAll(() => {
    // This is Mocha; in Jest, use beforeAll
    //shallow = createShallow()
    mount = createMount()
    wrapper = mount(
      <ThemeProvider theme={theme}>
        <UserSearch {...initialProps} />
      </ThemeProvider>
    )
    buttons = wrapper.find("button")
  })

  // what to do after each test
  afterAll(() => {
    //mount.cleanup()
    jest.clearAllMocks()
  })

  it("should work", () => {
    wrapper = mount(<UserSearch />)
  })

  // UI Integrity test
  it("should have submit buttons", () => {
    expect(buttons.length).toEqual(1)
  })
})

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}))
