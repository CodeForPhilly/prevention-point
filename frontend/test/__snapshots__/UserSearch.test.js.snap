// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSearch /> should match the snapshot 1`] = `"<div class=\\"MuiContainer-root participant-search MuiContainer-maxWidthLg\\"><div><h2 class=\\"MuiTypography-root participant-search__heading MuiTypography-h2 MuiTypography-displayBlock\\">Participant Search</h2><p class=\\"MuiTypography-root participant-search__copy MuiTypography-body1 MuiTypography-displayBlock\\"><b>Reminder:</b> Search for participant profile prior to creating a new profile</p><form class=\\"participant-search__form\\"><div class=\\"MuiFormGroup-root participant-search__input\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"user_id\\">User ID</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"user_id\\" name=\\"user_id\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><h2 class=\\"MuiTypography-root participant-search__heading MuiTypography-h2 MuiTypography-displayBlock\\">Or</h2><div class=\\"MuiFormGroup-root participant-search__input\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"first_name\\">First Name</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"first_name\\" name=\\"first_name\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><div class=\\"MuiFormGroup-root participant-search__input\\"><div class=\\"MuiFormControl-root\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated\\" data-shrink=\\"false\\" for=\\"last_name\\">Last Name</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"><input aria-invalid=\\"false\\" id=\\"last_name\\" name=\\"last_name\\" type=\\"text\\" class=\\"MuiInputBase-input MuiInput-input\\" value=\\"\\"></div></div></div><p class=\\"error-message\\"></p><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-appButton-226 MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled\\" tabindex=\\"-1\\" type=\\"submit\\" disabled=\\"\\"><span class=\\"MuiButton-label\\">Submit</span></button></form></div></div>"`;
